### PR TITLE
Add support for mysql to pgsql to DB migrator

### PR DESCRIPTION
**Pull request by wy65701436**
This commit is to implement db migration from mysql to pgsql, this is specific step for user to upgrade harbor across v1.5.0, as we have move harbor DB to pgsql from 1.5.0. It covers harbor db and notary db migration.

The upgrade path from v1.2.0 to v1.6.0;
v1.2.0(mysql) -> v1.5.0(mysql) -> v1.5.0(pgsql) -> v1.6.0(pgsql)

[Tested]
1, From 1.2.0 to current version(with and without notary)
2, From 1.3.0 to current version(with and without notary)
3, From 1.4.0 to current version(with and without notary)

[Usage]
1, upgrade harbor DB: docker run -it -e DB_USR=root -e DB_PWD=${pwd} -v /data/database/:/var/lib/mysql vmware/harbor-migrator:dev --db up
2, upgrade notary DB: docker run -it  -v /data/notary-db/:/var/lib/mysql -v /data/database:/var/lib/postgresql/data vmware/harbor-migrator:dev --db up

You need to run above two steps to migrate all DBs. If no require to migrate notary, just run step 1 only.

[Limitation]
For notary migration, there is a necessary precondition that you should not launch harbor before doing it, otherwise the migrator will give a waring message and stop the process.

[Where to get migrator]
From the local offline installer, this PR also fix issue #4847

### CODE REVIEW

File: tools/migration/db/run.sh [hunk 1]
```diff
@@ -1,125 +1,505 @@
...
+            ## dump mysql
+            mysqldump --compatible=postgresql --default-character-set=utf8 --databases registry > /harbor-migration/db/registry.mysql     
+            stop_mysql $DB_USR $DB_PWD
+
+            ## migrate 1.5.0-mysql to 1.5.0-pqsql.
+            python /harbor-migration/db/pgsql_migrator.py /harbor-migration/db/registry.mysql /harbor-migration/db/registry.pgsql
+
+            launch_pgsql $PGSQL_USR
+            psql -U $PGSQL_USR -f /harbor-migration/db/schema/registry_from_$cur_version.pgsql
+            psql -U $PGSQL_USR -f /harbor-migration/db/schema/notaryserver.pgsql
```
Comment (reasonerjt) [inline tools/migration/db/run.sh]: Why need to call notary pgsql in `up_harbor`?

File: tools/migration/db/run.sh [hunk 2]
```diff
@@ -1,125 +1,505 @@
...
+        fi
+    fi
+
+    # $cur_version <='1.5.0', $target_version >'1.5.0', it needs to upgrade to 1.5.0.mysql => 1.5.0.pgsql => target_version.pgsql.
+    if [ $v1_com != 1 ] && [ $v2_com = 1 ]; then
+        if [ $ISMYSQL != true ]; then
+            echo "Please make sure to mount the correct the data volumn."
+            return 1
+        else
+            alembic_up mysql '1.5.0' false
```
Comment (reasonerjt) [inline tools/migration/db/run.sh]: the third parm should be removed.

File: tools/migration/Dockerfile [hunk 3]
```diff
@@ -1,14 +1,43 @@
...
+    && chown -R mysql:mysql /var/lib/mysql /var/run/mysqld \
+    && chmod 777 /var/run/mysqld /docker-entrypoint-initdb.d /docker-entrypoint-updatedb.d \
     && tdnf install -y mariadb-devel python2 python2-devel python-pip gcc \
     linux-api-headers glibc-devel binutils zlib-devel openssl-devel \
     && pip install mysqlclient alembic \
-    && tdnf clean all \
-    && mkdir -p /harbor-migration
+    && mkdir -p /harbor-migration \
+    && touch /etc/localtime.bak \
+    && tdnf install -y sed shadow gzip postgresql >> /dev/null\
```
Comment (reasonerjt) [inline tools/migration/Dockerfile]: We should merge the `tdnf install`

File: tools/migration/Dockerfile [hunk 4]
```diff
@@ -1,14 +1,43 @@
...
+    && groupadd -r postgres --gid=999 \
+    && useradd -r -g postgres --uid=999 postgres \
+    && mkdir -p /docker-entrypoint-initdb.d \
```
Comment (reasonerjt) [inline tools/migration/Dockerfile]: This is not needed.

File: tools/migration/db/run.sh [hunk 5]
```diff
@@ -1,125 +1,505 @@
...
+    return 1
+}
+
+# This function is only for notary <= 1.5.0 mysql to pgsql.
+function up_notary {
+
+    set +e
+    chown -R 10000:10000 /var/lib/mysql
+    mysqld &
+    for i in {60..0}; do
```
Comment (reasonerjt) [inline tools/migration/db/run.sh]: Could we update `launch_mysql` to make it support different scenarios and pass an empty string as password?

File: tools/migration/db/pgsql_migrator.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+                insert_lines.append(line)
+                continue
+            else:    
+                insert_lines.append(line)
+    
+    write_database(pgsql_dump, "notarysigner")
+    write_insert(pgsql_dump, insert_lines)
+    write_sequence(pgsql_dump, "private_keys", "id")
+
+def drop_line(line):
```
Comment (reasonerjt) [inline tools/migration/db/pgsql_migrator.py]: it's not needed if all we wanna catch is `INSERT INTO` this method is not needed.

File: tools/migration/db/pgsql_migrator.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        if drop_line(line):
+            continue
+
+        # catch insert
+        if line.startswith("INSERT INTO"):
+            if line.find("\\") != -1:
+                line = line.replace('\\', '')
+                insert_lines.append(line)
+                continue
+            elif line.startswith('INSERT INTO "change_category"') | line.startswith('INSERT INTO "changefeed"') | line.startswith('INSERT INTO "schema_migrations"'):
```
Comment (reasonerjt) [inline tools/migration/db/pgsql_migrator.py]: We should not skip the `INSERT INTO` 'change_category' and 'changefeed'

File: tools/migration/db/pgsql_migrator.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+if __name__ == "__main__":
+    if sys.argv[1].find("registry") != -1:
+        do_parse_registry_db(sys.argv[1], sys.argv[2])
+    elif sys.argv[1].find("notaryserver") != -1:
+        do_parse_notary_server_db(sys.argv[1], sys.argv[2])
+    elif sys.argv[1].find("notarysigner") != -1:
+        do_parse_notary_signer_db(sys.argv[1], sys.argv[2])
+    else:
+        print ("Unsupport mysql dump file.")
```
Comment (reasonerjt) [inline tools/migration/db/pgsql_migrator.py]: Better print the sys.argv[1] into error message to provide more details.

File: tools/migration/db/pgsql_migrator.py (hunk 6 above)
Comment (reasonerjt) [inline tools/migration/db/pgsql_migrator.py]: This is not needed as we need only `INSERT INTO`

**Comment by reasonerjt** [comment]
Two overall comments:
1)  There should not be multiple pgsql script for different schema version, there should be only one for 1.5
2)  The run.sh should be refactored that there's one script to specifically handle mysql-> pgsql

**Comment by eymriseFrance** [comment]
Merci

File: tools/migration/db/util/alembic.sh [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    export PYTHONPATH=$PYTHONPATH:/harbor-migration/db/alembic
+    
+    if [ $db_type = "mysql" ]; then
+        source /harbor-migration/db/alembic/alembic.tpl > /harbor-migration/db/alembic/alembic.ini     
+        echo "Performing upgrade $target_version..."
+        alembic -c /harbor-migration/db/alembic/alembic.ini current
+        alembic -c /harbor-migration/db/alembic/alembic.ini upgrade $target_version
+        rc="$?"
+        alembic -c /harbor-migration/db/alembic/alembic.ini current
+    elif [ $db_type = "mysql" ]; then
```
Comment (reasonerjt) [inline tools/migration/db/util/alembic.sh]: This is still incorrect?

File: tools/migration/db/util/alembic.sh [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    local target_version="$2"
+
+    export PYTHONPATH=$PYTHONPATH:/harbor-migration/db/alembic
+    
+    if [ $db_type = "mysql" ]; then
+        source /harbor-migration/db/alembic/alembic.tpl > /harbor-migration/db/alembic/alembic.ini     
+        echo "Performing upgrade $target_version..."
+        alembic -c /harbor-migration/db/alembic/alembic.ini current
+        alembic -c /harbor-migration/db/alembic/alembic.ini upgrade $target_version
+        rc="$?"
```
Comment (reasonerjt) [inline tools/migration/db/util/alembic.sh]: how is `rc` used?

File: tools/migration/db/util/alembic.sh [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+source $PWD/db/util/pgsql.sh
+source $PWD/db/util/mysql.sh
```
Comment (reasonerjt) [inline tools/migration/db/util/alembic.sh]: Are these `source` needed?

File: tools/migration/db/run.sh [hunk 12]
```diff
@@ -1,125 +1,247 @@
...
+        exit 0
+    fi
+
+    init
+
+    local key="$1"
+
+    case $key in
+    up|upgrade)
+        upgrade
```
Comment (reasonerjt) [inline tools/migration/db/run.sh]: Where's the `$1`?

File: tools/migration/db/util/mysql_pgsql_1_5_0.sh [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+#
+
+source $PWD/db/util/pgsql.sh
+source $PWD/db/util/alembic.sh
+
+DBCNF="-hlocalhost -u${DB_USR}"
+
+function mysql_2_pgsql_1_5_0 {
+    
+    alembic_up mysql '1.5.0'
```
Comment (reasonerjt) [inline tools/migration/db/util/mysql_pgsql_1_5_0.sh]: what if this fails?

File: tools/migration/db/util/mysql_pgsql_1_5_0.sh [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+function up_notary {
+
+    set +e
+    if [[ $(mysql $DBCNF -N -s -e "select count(*) from information_schema.tables \
+                where table_schema='notaryserver' and table_name='tuf_files'") -eq 0 ]]; then
+        echo "no content trust data needs to be updated."
+        return 0
+    else
+
+        ## it's not a clean notary db, so cannot execute the create tables step.
```
Comment (reasonerjt) [inline tools/migration/db/util/mysql_pgsql_1_5_0.sh]: Doesn't Harbor registry DB has the same issue?

File: tools/migration/db/util/alembic.sh (hunk 9 above)
Comment (wy65701436) [reply tools/migration/db/util/alembic.sh]: will fix the typo.

File: tools/migration/db/util/alembic.sh (hunk 10 above)
Comment (wy65701436) [reply tools/migration/db/util/alembic.sh]: could be removed.

File: tools/migration/db/util/mysql_pgsql_1_5_0.sh (hunk 13 above)
Comment (wy65701436) [reply tools/migration/db/util/mysql_pgsql_1_5_0.sh]: the shell will be exited with non-zero code as "set -e"

File: tools/migration/db/util/mysql_pgsql_1_5_0.sh (hunk 14 above)
Comment (wy65701436) [reply tools/migration/db/util/mysql_pgsql_1_5_0.sh]: no, the init() handles the repeats run.

File: tools/migration/db/run.sh (hunk 12 above)
Comment (wy65701436) [reply tools/migration/db/run.sh]: at line 225.

File: tools/migration/db/run.sh [hunk 15]
```diff
@@ -1,125 +1,247 @@
...
+cur_version=""
+PGSQL_USR="postgres" 
+
+function init {
+    if [ "$(ls -A /var/lib/mysql)" ]; then
+        # As after the first success run, the data will be migrated to pgsql,
+        # the PG_VERSION should be in /var/lib/mysql if user repeats the UP command.
+        if [ -e '/var/lib/mysql/PG_VERSION' ]; then
+            ISPGSQL=true
+        elif [ "$(ls -A /var/lib/mysql/mysql)" ]; then
```
[the remark this answers is not in the document]
Comment (wy65701436) [reply tools/migration/db/run.sh]: -d is to check if a directory exists. 
ls -A is to check if a directory is empty.  

I believe later is better.

File: tools/migration/db/util/mysql_pgsql_1_5_0.sh (hunk 13 above)
Comment (wy65701436) [reply tools/migration/db/util/mysql_pgsql_1_5_0.sh]: will add set -e to each of these shell scripts.

File: tools/migration/db/util/alembic.sh (hunk 11 above)
Comment (wy65701436) [reply tools/migration/db/util/alembic.sh]: could be removed.

File: tools/migration/db/util/mysql_pgsql_data_converter.py [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+                                    'INSERT INTO "access_log" ("log_id", "username", "project_id", "repo_name", "repo_tag", "guid", "operation", "op_time")')
+                insert_lines.append(line)
+                continue
+            # pgsql doesn't support 0 as a time data, change it to the minimum value.
+            elif line.find("0000-00-00 00:00:00") != -1:
+                line = line.replace("0000-00-00 00:00:00", "0001-01-01 00:00:00")
+                insert_lines.append(line)
+                continue
+            # it's only for con_str, mysql dump the con_str to {\"kind\":\"Manual\",\"schedule_param\":null}.
+            elif line.find('INSERT INTO "replication_policy"') != -1:
```
Comment (reasonerjt) [inline tools/migration/db/util/mysql_pgsql_data_converter.py]: Please also consider the table `img_scan_overview`

**Pull request merged**